universal_bridge.md

```diff
@@ -0,0 +1,456 @@
+3. **L2↔L2 Bridging:** The bridge supports moving both **native ERC-20** and **CET** between OP L2s. 
+4. **L2↔L2 (ERC-20):** If the source asset is a **native ERC-20** on L2, it is **Locked** on the source L2 and a **CET is Minted** on the destination L2.  
+5. **L2↔L2 (CET):** If the source asset is a **CET**, it is **Burned** on the source L2 and **Minted** on the destination L2.  
+6. **L2→L1 Redemption:** Any CET can always be redeemed to **unlock the collateral on L1**.  
+7. **Proof System:** The design **builds on OP-Succinct** (validity proofs). L1 verifications rely on **per-L2 finalized post roots**
+8. **Mint Authority / Safety:** Only **bridge contracts** may mint or burn **CET**; no external mint paths.  
+9. **Exit Logic (proof-based paths):** Use the **same exit logic as OP-Succinct**: prove claims against a chain’s **post root**, then an **MPT/storage proof** to the chain’s **Outbox/Exit root**, then a **Merkle inclusion** for the exit record.  
+10. **Replay Protection:** Messages can be consumed only once. Any replay will be ignored.  
+11. **Inter-L2 Fast Path:** For **L2↔L2** transfers, the destination **mints on receipt of a bridge message** (no proof verification at claim time). **Later settlement** is done simultaneously via aggregated proofs (out of scope here).
+12. **TODO:** Allow token owner to have the bridge mint native token on specified conditions.
+
+
+___
+
+# L2↔L2 Bridge --- Sessioned Mailbox Flow
+
+
+### Entities & Contracts
+
+-  **ComposableERC20**
+    An ERC20 wrapper native to the bridge.
+
+-   **Bridge (per L2)**
+    Handles: locking native ERC-20, burning CET, mailbox write/read, and
+    minting (via token's `onlyBridge` gate when called from the bridge
+    context).
+
+-   **Mailbox (per L2)**
+    Minimal append-only message bus with read-once semantics per
+    `(chainSrc, recipient, sessionId, label)`.
+
+    -   `write(chainId, account, sessionId, label, payload)`\
+    -   `read(chainId, account, sessionId, label) → bytes`
+        (consumes/marks delivered)
+
+-   **Supported Token types**
+
+    -   **Native ERC-20** on an L2 (not minted by bridge).
+    -   **CET** (ComposeableERC20Token) --- canonical L2 representation
+        of L1/L2 asset.
+        -   `mint`/`burn` are **restricted** to `msg.sender == Bridge`.
+
+-   **Deterministic CET Addresses (Superchain-style)**
+    Each CET contract is deployed **at the same address across all OP
+    L2s**, deterministically derived from the L1 canonical asset address
+    (using CREATE2/CREATE3).
+
+    -   This eliminates the need for a per-chain registry.
+    -   Mailbox payloads only need to carry the **`remoteToken` address**.
+    -   On the destination chain, the bridge deterministically computes
+        the CET address and mints to the receiver.
+
+    **TODO** More details?
+
+------------------------------------------------------------------------
+
+##  ComposeableERC20
```

Comment:
It's an implementation detail, but to make the token IERC7802 compatible, we should implement:
```
function supportsInterface(bytes4 interfaceId) 
        external pure override returns (bool) 
    {
        return interfaceId == type(IERC7802).interfaceId ||
               interfaceId == type(IERC165).interfaceId;
    }
```